Microsoft.Toolkit.Uwp.Input.GazeControls/GazeFilePicker.xaml.cs

```diff
@@ -0,0 +1,390 @@
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="GazeFilePicker"/> class.
+        /// </summary>
+        public GazeFilePicker()
+        {
+            this.InitializeComponent();
+
+            this.Opened += OnGazeFilePickerOpened;
+
+            FileTypeFilter = new List<string>();
+
+            _initializationTimer = new DispatcherTimer();
+            _initializationTimer.Interval = TimeSpan.FromMilliseconds(125);
+            _initializationTimer.Tick += OnInitializationTimerTick;
+        }
+
+        private void OnInitializationTimerTick(object sender, object e)
+        {
+            _initializationTimer.Stop();
+            CreateFavoritesButtons();
+            CreateCommandSpaceButtons();
+            GazeKeyboard.Target = FilenameTextbox;
+            SetFileListingsLayout();
+            SetFileTypeFilterText();
+        }
+
+        private void SetFileTypeFilterText()
+        {
+            var filters = FileTypeFilter.Select(item => Regex.Replace(item, "^\u002E", "*."));
+            var allFilters = string.Join(", ", filters);
+            FileTypeFilterTextBlock.Text = allFilters;
+        }
+
+        private async void CreateFavoritesButtons()
+        {
+            var favoritesPanel = this.FindControl<StackPanel>("FavoritesPanel");
+            Debug.Assert(favoritesPanel != null, "KnownFoldersPanel not found");
+
+            var style = (Style)this.Resources["PickerButtonStyles"];
+
+            favoritesPanel.Children.Clear();
+
+            List<StorageFolder> favorites;
+            if ((Favorites == null) || (Favorites.Count == 0))
+            {
+                KnownFolderId[] knownFolderIds =
+                {
+                    KnownFolderId.DocumentsLibrary,
+                    KnownFolderId.PicturesLibrary,
+                    KnownFolderId.VideosLibrary,
+                    KnownFolderId.MusicLibrary
+                };
+
+                favorites = new List<StorageFolder>();
+                foreach (var folderId in knownFolderIds)
+                {
+                    try
+                    {
+                        var knownFolder = await KnownFolders.GetFolderAsync(folderId);
+                        favorites.Add(knownFolder);
+                    }
+                    catch (Exception)
+                    {
+                        continue;
+                    }
+                }
+            }
+            else
+            {
+                favorites = Favorites;
+            }
+
+            foreach (var folder in favorites)
+            {
+                var button = new Button();
+                button.Style = style;
+                button.Tag = folder;
+                button.Click += OnFavoritesClick;
+                button.Content = folder.Name;
+
+                favoritesPanel.Children.Add(button);
+            }
+        }
+
+        private void CreateCommandSpaceButtons()
```

Comment:
There's a lot of UI being created in code-behind, what is this trying to do?

Reply:
The default template for ContentDialog always insists on having a fixed number of buttons. For the File-Open control, this is sufficient. But for File..Save, I need additional buttons on the command bar for creating new folders and entering a filename. 

What is the right way to do this?

Reply:
If the layout is similar enough and it's just an extra button or two, but them in XAML and then just toggle their visibility. If the layouts are pretty different, then create two separate controls with the two different templates. You can create a common base class for both of them with any shared logic. FileOpen/SavePicker are usually separate in UWP today anyway.

Reply:
Yes, it is just an extra button that should be in the bottom row in the CommandSpace grid. I have tried customizing the template by adding a `    <Style TargetType="ContentDialog">` section to `<ContentDialog.Resources>` within the GazeFilePicker.xaml file. But the changes I make to the template don't have any affect. ( I added an additional button to the CommandSpace grid). 

I am pretty much doing the same thing in code. I am just recreating all the buttons for the sake of simplicity and readability. Is it okay to retain this as is?

Reply:
In the latest commit, I have refactored the picker into separate GazeFileOpenPicker and GazeFileSavePicker. I have also moved the controls creation to XAML.